src/backend/utils/resgroup/resgroup.c

```diff
@@ -4617,38 +4622,75 @@ HandleMoveResourceGroup(void)
 	ResGroupData *group;
 	ResGroupData *oldGroup;
 
+	Assert(Gp_role == GP_ROLE_DISPATCH || Gp_role == GP_ROLE_EXECUTE);
+
 	/* transaction has finished */
-	if (!selfIsAssigned())
+	if (!IsTransactionState() || !selfIsAssigned())
+	{
+		SpinLockAcquire(&MyProc->movetoMutex);
+		/* notify initiator if command still relevant */
+		if (MyProc->movetoResSlot)
+			(void) ResGroupMoveNotifyInitiator(MyProc->movetoCallerPid);
```

Comment:
Is this operation short enough to hold spinlock?

Reply:
With this simplified version, we need to do `ResGroupMoveNotifyInitiator()` and `MyProc->movetoResSlot = NULL` inside one lock or initiator may read incorrect MyProc->movetoResSlot result.